pkg/proxy/runner/bounded_frequency_runner.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package runner
+
+import (
+	"time"
+
+	"k8s.io/utils/clock"
+)
+
+type BoundedFrequencyRunner struct {
+	maxInterval time.Duration
+	minInterval time.Duration
+	fn          func()
+	queue       chan struct{}
+
+	clock            clock.Clock
+	maxIntervalTimer clock.Timer
+	minIntervalTimer clock.Timer
+}
+
+// NewBoundedFrequencyRunner creates and returns a new BoundedFrequencyRunner.
+// This runner manages the execution frequency of the provided function `fn`.
+//
+// The runner guarantees two properties:
+//  1. Minimum Interval (`minInterval`): At least `minInterval` must pass between
+//     the *completion* of one execution and the *start* of the next. Calls to
+//     `Run()` during this cooldown period are coalesced and deferred until the
+//     interval expires. This prevents burst executions.
+//  2. Maximum Interval (`maxInterval`): The function `fn` is guaranteed to run
+//     at least once per `maxInterval`, ensuring periodic execution even without
+//     explicit `Run()` calls (e.g., for refreshing state).
+//
+// Internally, the runner uses a buffered channel with a capacity of 1, acting
+// as a single-item queue. A dedicated goroutine (started by `Loop`) consumes
+// from this queue. This consumer loop triggers an execution of `fn` under two
+// conditions:
+//  1. An item is received from the queue (signaled by a `Run()` call).
+//  2. The `maxInterval` timer expires.
+//
+// After `fn` completes, the loop waits for the `minInterval` timer to expire
+// before it becomes eligible to execute `fn` again. This ensures the minimum
+// time between executions.
+//
+// Calls to `Run()` attempt to add an item to the queue. If the queue is already
+// full (meaning an execution is already pending), the `Run()` call is effectively
+// ignored (coalesced), as the existing item in the queue guarantees a future
+// execution.
+//
+// `maxInterval` must be greater than or equal to `minInterval`; otherwise,
+// this function will panic.
+func NewBoundedFrequencyRunner(fn func(), minInterval, maxInterval time.Duration) *BoundedFrequencyRunner {
```

Comment:
It would be helpful to have one commit that just copies BFR over to `pkg/proxy/runner` unchanged, then a second commit to get to this state, to make it clearer what is the same and what is different from the old code.